pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,441 @@
+		pvList, apiError = c.commonClient.KubeInformers.InformersFor("").Core().V1().PersistentVolumes().Lister().List(labels.Everything())
+		if apiError != nil {
+			klog.Errorf("error listing volumes objects: %v", apiError)
+			return false, nil
+		}
+		return true, nil
+	})
+	return pvList, err
+}
+
+func (c *EBSVolumeTagsController) updateVolumeWithRetry(ctx context.Context, pv *v1.PersistentVolume) error {
+	backoff := wait.Backoff{
+		Duration: operationDelay,
+		Factor:   operationBackoffFactor,
+		Steps:    operationRetryCount,
+	}
+	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		var apiError error
+		_, apiError = c.commonClient.KubeClient.CoreV1().PersistentVolumes().Update(ctx, pv, metav1.UpdateOptions{})
+		if apiError != nil {
+			klog.Errorf("error updating volume object %s: %v", pv.Name, apiError)
+			return false, nil
+		}
+		return true, nil
+	})
+	return err
+}
+
+func (c *EBSVolumeTagsController) handleTagUpdateFailure(batch []*v1.PersistentVolume, error error) {
+	errorMessage := fmt.Sprintf("Error updating tags for volume %v: %v", batch, error)
+	for _, pv := range batch {
+		klog.Errorf("Error updating volume %v tags: %v", pv.Name, errorMessage)
+		c.failedQueue.AddRateLimited(pv.Name)
+	}
+	var pvNames []string
+	for _, pv := range batch {
+		pvNames = append(pvNames, pv.Name)
+	}
+	// Emit a warning event for the failure
+	c.eventRecorder.Warning("EBSVolumeTagsUpdateFailed", fmt.Sprintf("Failed to update tags for batch %v: %v", pvNames, errorMessage))
+}
+
+// newAndUpdatedTags adds and update existing AWS tags with new resource tags from OpenShift infrastructure
+func newAndUpdatedTags(resourceTags []configv1.AWSResourceTag) []*ec2.Tag {
+	// Convert map back to slice of ec2.Tag
+	var tags []*ec2.Tag
+	for _, tag := range resourceTags {
+		tags = append(tags, &ec2.Tag{
+			Key:   aws.String(tag.Key),
+			Value: aws.String(tag.Value),
+		})
+	}
+	return tags
+}
+
+func filterUpdatableVolumes(volumes []*v1.PersistentVolume, newTagsHash string) []*v1.PersistentVolume {
+	var pvsToBeUpdated = make([]*v1.PersistentVolume, 0)
+	for _, volume := range volumes {
+		if volume.Spec.CSI != nil && volume.Spec.CSI.Driver == driverName {
+			existingHash := getPVTagHash(volume)
+			if existingHash == "" || existingHash != newTagsHash {
+				pvsToBeUpdated = append(pvsToBeUpdated, volume)
+			}
+		}
+	}
+	return pvsToBeUpdated
+}
+
+func pvsToResourceIDs(volumes []*v1.PersistentVolume) []*string {
+	var resourceIDs []*string
+	for _, volume := range volumes {
+		resourceIDs = append(resourceIDs, aws.String(volume.Spec.CSI.VolumeHandle))
+	}
+	return resourceIDs
+}
+
+// setPVTagHash stores the hash in the PV annotations.
+func setPVTagHash(pv *v1.PersistentVolume, hash string) {
+	// Ensure the PV has an annotations map
+	if pv.Annotations == nil {
+		pv.Annotations = make(map[string]string)
+	}
+
+	// Set or update the tag hash annotation
+	pv.Annotations[tagHashAnnotationKey] = hash
```

Comment:
We should get `DeepCopy` before modifying object we retrieved from informer. We do not want to modify object we retrieved from informers directly.